integreat_cms/cms/views/pages/partial_page_tree_view.py

```diff
@@ -1,61 +1,74 @@
-from django.shortcuts import render
-from django.utils.decorators import method_decorator
-from django.views.generic import TemplateView
+from collections import defaultdict
+import json
 
-from integreat_cms.cms.views.pages.page_context_mixin import PageContextMixin
+from django.http import JsonResponse
+from django.template.loader import render_to_string
+from django.utils.translation import get_language
+from django.views.decorators.http import require_POST
 
+from ..pages.page_context_mixin import PageContextMixin
+from ...models.languages.language import Language
 from ...decorators import permission_required
 
 
-@method_decorator(permission_required("cms.view_page"), name="dispatch")
-class PartialPageTreeView(TemplateView, PageContextMixin):
-    """
-    View for rendering a partial page tree
-    """
+@permission_required("cms.view_page")
+@require_POST
+# pylint: disable=unused-argument
+def render_partial_page_tree_views(request, region_slug, language_slug):
+    r"""
+    Retrieve the rendered subtree of a given root page
 
-    #: Template for a partial page tree
-    template = "pages/_page_tree_children.html"
+    :param request: The current request
+    :type request: ~django.http.HttpRequest
 
-    # pylint: disable=unused-argument
-    def get(self, request, *args, **kwargs):
-        r"""
-        Retrieve the rendered subtree of a given root page
+    :param region_slug: The slug of the current region
+    :type region_slug: str
 
-        :param request: The current request
-        :type request: ~django.http.HttpRequest
+    :param language_slug: The slug of the current language
+    :type language_slug: str
 
-        :param \*args: The supplied arguments
-        :type \*args: list
+    :return: The rendered template responses
+    :rtype: ~django.http.JsonResponse
+    """
+    requested_tree_ids = [int(i) for i in json.loads(request.body.decode("utf-8"))]
 
-        :param \**kwargs: The supplied keyword arguments
-        :type \**kwargs: dict
+    region = request.region
+    language = region.get_language_or_404(language_slug, only_active=True)
 
-        :return: The rendered template response
-        :rtype: ~django.template.response.TemplateResponse
-        """
-        tree_id = int(kwargs.get("tree_id"))
-        region = request.region
-        language = region.get_language_or_404(
-            kwargs.get("language_slug"), only_active=True
-        )
+    backend_language = Language.objects.filter(slug=get_language()).first()
+
+    all_pages = (
+        region.pages.filter(tree_id__in=requested_tree_ids)
+        .prefetch_major_public_translations()
+        .prefetch_related("mirroring_pages")
+        .cache_tree(False)
+    )
+
+    pages_by_id = defaultdict(list)
+    for page in all_pages:
+        pages_by_id[page.tree_id].append(page)
```

Comment:
```suggestion
    pages_by_id = {page.tree_id: page for page in all_pages}
```
Using a `defaultdict` only makes sense when it actually prevents errors - in this case however it does only delay it.
`pages = pages_by_id[tree_id]` does not throw a `KeyError`, but the next statement `parent = pages[0]` does if the default dict returns an empty list.
So probably it's best to go with boring old dictionaries here.

Reply:
The idea behind this code is that `all_pages` contains all pages that are requested, but they still need to get grouped according to their trees, so that they can be sent to the client. I think this suggestion changes the behavior: previously a page tree could contain a list of pages, but with this change only a single page.

Reply:
Ahh sorry! So the defaultdict should not prevent the key error, but make it easier to directly append the pages instead of checking whether the `tree_id` is already contained in `pages_by_id`? Yes makes sense! :+1: